fxa-rust-client/src/lib.rs

```diff
@@ -0,0 +1,502 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// For error_chain:
+#![recursion_limit = "128"]
+
+extern crate base64;
+#[macro_use]
+extern crate error_chain;
+extern crate hawk;
+extern crate hex;
+extern crate hkdf;
+extern crate hmac;
+extern crate jose;
+#[macro_use]
+extern crate lazy_static;
+#[macro_use]
+extern crate log;
+extern crate openssl;
+extern crate rand;
+extern crate regex;
+extern crate reqwest;
+extern crate serde;
+#[macro_use]
+extern crate serde_derive;
+#[macro_use]
+extern crate serde_json;
+extern crate sha2;
+extern crate url;
+
+use std::collections::HashMap;
+use std::mem;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use self::login_sm::LoginState::*;
+use self::login_sm::*;
+use errors::*;
+use http_client::browser_id::jwt_utils;
+use http_client::{Client, OAuthTokenResponse, ProfileResponse};
+use jose::{JWKECCurve, JWE, JWK};
+use openssl::hash::{hash, MessageDigest};
+use rand::{OsRng, RngCore};
+use url::Url;
+use util::now;
+
+mod config;
+pub mod errors;
+pub mod http_client;
+mod login_sm;
+mod oauth;
+mod util;
+
+pub use config::Config;
+
+// If a cached token has less than `OAUTH_MIN_TIME_LEFT` seconds left to live,
+// it will be considered already expired.
+const OAUTH_MIN_TIME_LEFT: u64 = 60;
+// A cached profile response is considered fresh for `PROFILE_FRESHNESS_THRESHOLD` ms.
+const PROFILE_FRESHNESS_THRESHOLD: u64 = 120000; // 2 minutes
+
+#[derive(Clone, Serialize, Deserialize)]
+struct StateV1 {
+    client_id: String,
+    config: Config,
+    login_state: LoginState,
+    oauth_cache: HashMap<String, OAuthInfo>,
```

Comment:
Is it accurate that the contents of `oauth_cache` are serialized and deserialized as part of the state of this object, and hence will persist across application restarts?

Assuming so, that's good, but it means that the name "cache" here is not really accurate.  The `oauth_cache` dict contains a mix of long-lived data persistent data (refresh tokens, keys) and short-lived transient data (access tokens).  I wonder if it's worth splitting them up into a persistent `oauth_tokens` map that contains the long-lived data and gets round-tripped through serialization, and a transient `oauth_cache` map that contains the short-lived data and is discarded when you serialize.